### DISCUSSION TITLE
Перевод FHIR на русский язык

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#252] [Минимальный набор артефактов]
При публикации RuCore и дочерних руководств проявляются фрагменты англоязычной документации.
<img width="692" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/0f1d9230-e052-4a7d-b7a5-45cc80209c2b">
Часть документации будет англоязычной всегда, например наименования атрибутов, но другая часть могла бы бы стать руссокязычной, если бы мы наследовали свои руководства не от базового руководства Fhir, а от переведенного.

Над переводом в свое время работали сотрудники Health Samurai, вот комментарии Александры Павлышиной @Semalexa 

- Да, в прошлом такая работа велась, сайт стандарта версии 3.0.1 был практически полностью переведен, можно посмотреть архивную версию - https://web.archive.org/web/20171010001134/http://fhir-ru.github.io/documentation.html.
После выхода четвертой версии была попытка обновить перевод сайта, для этого надо было постранично находить разницу между старой и новой англ. версией и вносить эти изменения в рус. перевод, что тоже в какой-то мере было проделано, но возникли технические сложности, сайт с переводом не билдился, билдился только с англ. версией.
- Технически наш сайт с переводом - это такой же сайт, как сайт стандарта (во всяком случае так было на тот момент), то есть он полноценно валидируется и собирается со всеми проверками ресурсов, примеров и т.д. И если допущена какая-то ошибка в файле, где делается перевод, то сайт не собирается.
- В том виде, как это было реализовано, работа по переводу и поддержке сайта была очень трудоемкой и технически сложной, так как надо было поддерживать возможность сверять диффы предыдущей англ. и новой версии.

В 2023 году в рамках подготовки материалов для ГОСТ по "персональным медицинским помощникам" И.В.Емелин сделал перевод спецификаций FHIR R5.0.0 в части REST API, поиска, типов данных, а также набора ресурсов, относящихся к моделированию результатов измерений. Результат представляет собой текст в формате MS Word с диаграммами и таблицами. Текст будет доступен публично после утверждения ГОСТ, мы также работаем над возможностью официального получения предварительных версий.

По мнению А.Павлышиной, этот текст может быть использован для создания русскоязычной спецификации FHIR в виде стандартного руководства по реализации, однако объем работы все равно представляется большим.

**Comment by Vrodvi** [comment]
К сожалению утверждения про объём перевода представлены субъективно.
Если бы смогли более формально представить сообществу объём, подлежащий переводу, то можно было бы что-то планировать, распараллеливать работу по переводу и отслеживать прогресс.

Сама спецификация представляет из себя html-сайт в Git репозитории (https://github.com/HL7/fhir/tree/master/source) - исходные страницы в репозитории легко находятся, к примеру:
<img width="144" alt="image" src="https://github.com/fhir-ru/core/assets/92928137/79fe6c99-5511-436b-941a-9319c54b840b">

Есть предложение составить "карту" сайта спецификации FHIR R5 с перечнем всех страниц, составляющих его (в идеале с осмысленной группировкой, вроде той что использована в навигации сайта спецификации - насколько это возможно для гипертекстового формата, т.к. очевидно, что на одну страницу можно выйти из множества мест), в виде плоского перечня файлов, требующих перевода. 
Это будет уже первый шаг для декомпозиции работы по переводу.
Если получится пробежаться по каждому файлу и посчитать в нём количество слов, то оценку объёма можно будет ещё повысить (слова разметки и прочую сопроводиловку, не требующую перевода, исключить из подсчёта по возможности), т.к. некоторые страницы будет существенно объёмней прочих (к примеру страница с описанием REST API достаточно объёмна сама по себе).

**Reply by Vrodvi** [reply]
@OlegPenzin

**Comment by Vrodvi** [comment]
Использование некого ГОСТа, который не известно в какой мере будет соответствовать базовой спецификации и не известно в каком объёме эту спецификацию будет покрывать, который придётся вручную куда-то переносить, выглядит как очень много дублирующей работы.

<РЕЖИМ НАИВНОСТИ: ВКЛЮЧЁН>
Нет ли возможности попробовать организовать эту работу в обратную сторону - сначала выбрать те разделы базовой спецификации, которые планируется сделать частью ГОСТа, перевести их в базовой спецификации и создать скрипты/утилиты для конвертации уже переведённого наполнения базовой спецификации в другой формат - в формат вордовских документов с форматированием под ГОСТ и всей необходимой сопутствующей трансформацией (ну или на худой конец копировать для ГОСТа фрагменты вручную, но будучи уверенными, что наполнение ГОСТа действительно будет соответствовать спецификации)?
<РЕЖИМ НАИВНОСТИ: ВЫКЛЮЧЕН>

**Comment by eugenekogan-spb** [comment]
Есть еще вопросы к ранее сделанному переводу:
- использовались html - тексты страниц или какие-то лежащие под ними исходники?
- как происходил процесс перевода - где лежали какие файлы, как была организована работа?
- почему в точности перевод не смог продолжиться для R4 и какие меры предосторожности следует принять, чтобы это не повторилось теперь?
@Semalexa

**Reply by Semalexa** [reply]
В основном все файлы, подлежащие переводу, находятся в папке source. 
В корне папки находятся .html-файлы. 
Каждому типу ресурса соответствует одноименная папка, в которой находится стандартный для ресурса набор файлов.
И есть файлы c java-кодом, где нужно переводить заголовки, названия колонок таблиц и т.п.

**Reply by Semalexa** [reply]
### Список файлов из корня папки source

administration-module.html
ansi.html
ansi-conformance.html
ansi-infrastructure.html
ansi-observation.html
ansi-patient.html
async.html
backboneelement.html
backboneelement-definitions.html
ballot-intro.html
best-practices.html
bindings.ini
cda-intro.html
change.html
clinicalreasoning-cds-on-fhir.html
clinicalreasoning-knowledge-artifact-distribution.html
clinicalreasoning-knowledge-artifact-representation.html
clinicalreasoning-module.html
clinicalreasoning-quality-reporting.html
clinicalreasoning-topics-definitional-resources.html
clinicalreasoning-topics-supporting-documentation.html
clinicalreasoning-topics-template.html
clinicalreasoning-topics-using-expressions.html
clinicalsummary-module.html
comparison.html
comparison-cda.html
comparison-other.html
comparison-v2.html
comparison-v3.html
compartments.xml
conformance-module.html
conformance-rules.html
cpt.html
credits.html
cvx.html
datatypes.html
datatypes-definitions.html
datatypes-examples.html
datatypes-extras.html
datatypes-mappings.html
datatypes-version-maps.html
defining-extensions.html
definitions.xsd
diagnostics-module.html
diff.html
directory.html
documentation.html
documents.html
dosage.html
dosage-definitions.html
dosage-examples.html
dosage-mappings.html
downloads.html
ehr-fm.html
element.html
elementdefinition.html
elementdefinition-definitions.html
elementdefinition-examples.html
elementdefinition-extras.html
elementdefinition-mappings.html
element-definitions.html
element-extras.html
exchange-module.html
extensibility.html
extensibility-definitions.html
extensibility-examples.html
extensibility-registry.html
fhir.css
fhir.ini
fhir-all.xsd
fhirpatch.html
fhirpath.html
fhirpath.md
fhir-print.css
fhir-xhtml.xsd
financial-module.html
fmg.html
formats.html
foundation-module.html
genomics.html
glossary.html
graphql.html
help.html
hierarchy.xml
history.html
http.html
icd.html
identifier-registry.html
igs.xml
implsupport-module.html
index.html
integrated-examples.html
iso3166.html
json.html
languages.html
license.html
lifecycle.html
logical.html
loinc.html
managing.html
mapping-language.html
mappings.html
mappingSpaces.xml
mapping-tutorial.html
marketingstatus.html
marketingstatus-definitions.html
marketingstatus-examples.html
marketingstatus-mappings.html
mdc.html
medications-module.html
messaging.html
metadatatypes.html
metadatatypes-definitions.html
metadatatypes-examples.html
metadatatypes-mappings.html
modules.html
modules-fragment.html
modules-list.html
narrative.html
narrative-definitions.html
narrative-example.html
navigation.xml
ncimeta.html
ndc.html
ndfrt.html
nd-json.html
newfooter.html
newheader.html
newnavbar.html
ns.html
ontology.xlsx
operations.html
operationslist.html
op-example-request.html
overview.html
overview-arch.html
overview-clinical.html
overview-dev.html
page.html
population.html
population-definitions.html
population-examples.html
population-mappings.html
prodcharacteristic.html
prodcharacteristic-definitions.html
prodcharacteristic-examples.html
prodcharacteristic-mappings.html
productshelflife.html
productshelflife-definitions.html
productshelflife-examples.html
productshelflife-mappings.html
profilelist.html
profiling.html
profiling-examples.html
pushpull.html
qa.html
r2maps.html
r3maps.html
rdf.html
redirect.asp
redirect.cgi
redirect.html
redirect.php
references.html
references-definitions.html
resourceguide.html
resourcelist.html
resourcelist-examples.html
resources.plantuml
resources.svg
resources-definitions.html
resources-examples.html
rxnorm.html
safety.html
sc.html
search.html
search_filter.html
searchparameter-registry.html
secpriv-module.html
security.html
security-labels.html
services.html
signatures.html
snomedct.html
snomedct-usage.html
sparql-rules.xml
status-codes.xml
storage.html
substanceamount.html
substanceamount-definitions.html
substanceamount-examples.html
substanceamount-mappings.html
summary.html
suppressed-workflow-warnings.xml
template.html
template-abstract.html
template-abstract-book.html
template-abstract-definitions.html
template-abstract-examples.html
template-book.html
template-book-defn.html
template-book-ex.html
template-comparison.html
template-comparison-set.html
template-compartment.html
template-compartment-book.html
template-conformance-pack.html
template-cs.html
template-cs-book.html
template-cs-ig.html
template-cs-ig-book.html
template-definitions.html
template-dependencies.html
template-dictionary.html
template-example.html
template-example-json.html
template-examples.html
template-example-shex.html
template-example-ttl.html
template-example-xml.html
template-explanations.html
template-extension.html
template-extension-definitions.html
template-extension-mappings.html
template-ig-operations.html
template-ig-toc.html
template-json-schema.html
template-logical.html
template-logical-definitions.html
template-logical-examples.html
template-logical-mappings.html
template-mappings.html
template-operation.html
template-operations.html
template-print.html
template-profile.html
template-profile-constraint.html
template-profile-definitions.html
template-profile-example-json.html
template-profile-examples.html
template-profile-example-xml.html
template-profile-mappings.html
template-profile-questionnaire.html
template-profiles.html
template-questionnaire.html
template-search-parameter.html
template-status-map.html
template-tx.html
template-tx-book.html
template-version-maps.html
template-vs.html
template-vs-book.html
template-vs-ig.html
template-vs-ig-book.html
terminologies.html
terminologies-binding-examples.html
terminologies-bindings.html
terminologies-conceptmaps.html
terminologies-systems.html
terminologies-valuesets.html
terminology-module.html
terminology-service.html
test-cases.xml
testing.html
toc.html
todo.html
ucum.html
ucum-essence.xml
uml-example.plantuml
uml-example.svg
unii.html
updates.html
usecases.html
validation.html
versioning.html
versions.html
w5.html
w5.ini
wglist.html
workflow.html
workflow-ad-hoc.html
workflow-communications.html
workflow-examples.html
workflow-management.html
workflow-module.html
xml.html
xml.xsd
xmldsig-core-schema.xsd

**Comment by Semalexa** [comment]
Для каждого ресурса в основном нужно переводить три файла: spreadsheet, introduction, notes.
Например для ресурса account это файлы:

- account-introduction.xml
- account-notes.xml
- account-spreadsheet.xml

**Comment by Semalexa** [comment]
Алгоритм обновления сайта FHIR (до версии R4.0.1):

https://dev.azure.com/fhir-build/fhir-ru/_build?definitionId=5&_a=summary&view=runs

0) git pull
1) скопировать папки guides.en, guides (рус. перевод), implementations, tools, source и source.ru из проекта в другое место
2) $ git fetch fhir-svn
3) $ git merge fhir-svn/R4Final
или если нужен определённый коммит, то
git merge <commitSHA>
4) resolve конфликты   
git checkout --theirs tools/
ga [conflicted files]   
git commit
5) проверить еще раз git merge fhir-svn/master - должно быть Already up-to-date.
6) получить список изменений в папке source 
git diff --name-status HEAD^ HEAD > diff-12-04-2020.txt
8) добавить новые файлы 
rsync --ignore-existing -r -v source/ source.ru/
9) удалить устаревшие файлы
rsync --ignore-existing --delete -r -v source/ source.ru/
9.1) строка ниже выводит список файлов с ру переводом, которые нужно будет, возможно, смержить (правая колонка в экселе)
перейти в папку source.ru
cd source.ru
выполнить:
grep -Hr [а-я] | cut -d: -f1 | sort -u  > filesWithRusTranslation-10-01-2021.txt
cd ..
9.2) открыть diff, оставить только M source/
заменить "deleted:   .+$"    "renamed:   .+$"     "new file:   .+$"     и   "^\s+\r\n" and "^\s+\n" и "	modified:   source/"  или "^D\s+.+$" и "^A\s+.+$" и "^M	source/"  - пустые строки на пустоту
9.3) в экселе две колонки - слева все модифицированные файлы из папки source/, справа список файлов с ру переводом, подсветить в левой колонке те файлы, которые есть в правой колонке формулой =NOT(ISNA(MATCH(RC1,C2,0))), остальные (белые незакрашенные ячейки - просто скопировать в source.ru/ командой
перед этим положить файл files-to-copy.txt в корень проекта

rsync -av --files-from=files-to-copy.txt source/ source.ru/
а подсвеченные - смержить 
10) spreadsheets смержить с помощью утилиты spreadsheet_parser.rb
для этого:
- взять из списка files-to-merge все spreadsheet.xml и datatypes/*.xml, terminologies/bindings.xml,  и положить их в files-to-copy.txt
для этого заменить:
(.+-spreadsheet.xml)
* $1
сортировать
удалить "* "
- создать копию папки source.ru и назвать ее source.ru-RU
- скопировать англ версию файлов в папку с ру переводом source.ru командой
rsync -av --files-from=files-to-copy.txt source/ source.ru/
- все spreadsheet.xml положить в files-to-parse.txt + 1 пустая строка в конце
- выполнить команду через обычный терминал cmd из корня fhir-svn:
ruby translation_parser.rb files-to-parse.txt > result_10-01-2021.txt
(gem install colorize) - для цветного вывода программы
11) выполнить мерж переводов с помощью winmerge: открыть сравнение папок source/ и source-old/, и сравнение папок source/ и source.ru/
11.1) при необходимости воспользоваться утилитой для мержа перевода спредшита пофайлово: источник-файл с переводом-оригиналом должен лежать в корне папки source.ru-RU, дестинейшен = путь к файлу от папки source.ru
ruby spreadsheet_translation_parser.rb guidanceresponse-spreadsheet.xml plandefinition\plandefinition-spreadsheet.xml
12) поменять версию сборки в файле 
открыть проводник из консоли в текущей папке
\fhir-svn\tools\java\org.hl7.fhir.tools.core\src\org\hl7\fhir\tools\publisher\Publisher.java
cd tools/java/org.hl7.fhir.tools.core/src/org/hl7/fhir/tools/publisher
np Publisher.java
line 488
 return "9663";
 return "10521";
 return "11828";
 return "12545";
 13) поменять дату копирования, ссылку на англ.
\fhir-svn\source.ru\newnavbar.html

\fhir-svn\tools\java\org.hl7.fhir.tools.core\src\org\hl7\fhir\tools\publisher\PageProcessor.java
  public final static String CI_PUB_NOTICE = 
      "<p style=\"background-color: gold; border:1px solid maroon; padding: 5px;\">\r\n"+
          "Это непрерывная интеграционная сборка FHIR (может быть некорректной время от времени). См. <a href=\"http://hl7.org/fhir/directory.html\">указатель опубликованных версий</a>\r\n"+
          "</p>\r\n";
            public final static String CI_PUB_NOTICE = 
      "<p style=\"background-color: gold; border:1px solid maroon; padding: 5px;\">\r\n"+
          "Это предварительный выпуск будущей версии FHIR (предположительно STU 3). Это версия STU 3 для голосования, а также для Коннектатона в Балтиморе. Текущей версией FHIR является <a href=\"http://hl7.org/fhir/index.html\">DSTU 2</a>. См. также <a href=\"http://hl7.org/fhir/directory.html\">Список всех доступных версий</a>.\r\n"+
          "</p>\r\n";

**Comment by Semalexa** [comment]
> * почему в точности перевод не смог продолжиться для R4 и какие меры предосторожности следует принять, чтобы это не повторилось теперь?
>   @Semalexa

Если задача просто взять сайт спецификации на какой-то момент времени и перевести его, то проблем не должно возникнуть.
Нужно просто взять желаемую версию спецификации и перевести файлы в папке source и еще в нескольких других местах.
Сбилдить сайт и опубликовать сгенерированую сборку.

Проблемы возникали, когда нужно было на уже переведенную версию поставить обновление спецификации. Для этого нужно было сначала понять, что поменялось в англ. версии, потом найти это место в русском переводе и внести правки.

Также проблема в том, что сборка занимает долгое время. Не посмотреть результаты перевода быстро.

Другая проблема, что если допущена какая-то ошибка в файле при переводе (непарный тег в html/xml, непарные кавычки и т.п.), то сайт не сбилдится.

**Comment by Semalexa** [comment]
В процессе последнего обновления переводимой версии спецификации сайт перестал билдиться (сборка настроена на папку source.ru, где должны лежать переведенные файлы), поэтому временно туда были скопированы оригинальные файлы без перевода, чтобы сайт сбилдился.
